src/Servers/Kestrel/test/InMemory.FunctionalTests/Http3/Http3TestBase.cs

```diff
@@ -143,385 +158,193 @@ protected void CreateConnection()
-
-                _pair = DuplexPipe.CreateConnectionPair(inputPipeOptions, outputPipeOptions);
-                
-                StreamContext = new TestStreamContext(canRead: true, canWrite: true, _pair, this);
-            }
-
-            public async Task<bool> SendHeadersAsync(IEnumerable<KeyValuePair<string, string>> headers, bool endStream = false)
-            {
-                var outputWriter = _pair.Application.Output;
-                var frame = new Http3RawFrame();
-                frame.PrepareHeaders();
-                var buffer = _headerEncodingBuffer.AsMemory();
-                var done = _qpackEncoder.BeginEncode(headers, buffer.Span, out var length);
-                frame.Length = length;
-                // TODO may want to modify behavior of input frames to mock different client behavior (client can send anything).
-                Http3FrameWriter.WriteHeader(frame, outputWriter);
-                await SendAsync(buffer.Span.Slice(0, length));
-
-                if (endStream)
+                if (streamType == ControlStreamId)
                 {
-                    await _pair.Application.Output.CompleteAsync();
+                    _inboundControlStream = stream;
+                    _controlStreamCreated.SetResult(null);
                 }
-
-                return done;
-            }
-
-            internal async Task SendDataAsync(Memory<byte> data, bool endStream = false)
-            {
-                var outputWriter = _pair.Application.Output;
-                var frame = new Http3RawFrame();
-                frame.PrepareData();
-                frame.Length = data.Length;
-                Http3FrameWriter.WriteHeader(frame, outputWriter);
-                await SendAsync(data.Span);
-
-                if (endStream)
+                else if (streamType == EncoderStreamId)
                 {
-                    await _pair.Application.Output.CompleteAsync();
+                    _inboundEncoderStream = stream;
+                }
+                else if (streamType == DecoderStreamId)
+                {
+                    _inboundDecoderStream = stream;
                 }
             }
+        }
 
-            internal async Task<Dictionary<string, string>> ExpectHeadersAsync()
-            {
-                var http3WithPayload = await ReceiveFrameAsync();
-                _qpackDecoder.Decode(http3WithPayload.PayloadSequence, this);
-                return _decodedHeaders;
-            }
-
-            internal async Task<Memory<byte>> ExpectDataAsync()
-            {
-                var http3WithPayload = await ReceiveFrameAsync();
-                return http3WithPayload.Payload;
-            }
-
-            internal async Task<Http3FrameWithPayload> ReceiveFrameAsync(uint maxFrameSize = Http3PeerSettings.DefaultMaxFrameSize)
+        private async ValueTask<long> TryReadStreamIdAsync(TestHttp3ControlStream controlStream)
```

Comment:
So this was copied from Http3ControlStream? That's normally what should be doing this, right? I'm worried we're testing too much of Http3TestBase and too little of the actual code.

Reply:
Http3ControlStream only is testing the server side element of a stream. The client still needs to be able to read the stream id and set the correct HTTP/3 unidirectional streams.